Fix Adding multiple topology constraints for osd prepare pod

kesavanvt opened the pull request:
Fixes : #879 

Signed-off-by: kesavan kvellalo@redhat.com

jarrpa commented:
I odn't quite understand what's going on... are you adding the same TSC twice? What does that accomplish?

kesavanvt commented:
> I odn't quite understand what's going on... are you adding the same TSC twice? What does that accomplish?

Basically, i have added twice because one TSC will act for zone/rack based on the runtime calculation (the topology key will be set at runtime) and the other would be for host by defaultwhich makes OSDs to distribute equally among the zone/rack. In case of flexible scaling, only one TSC will be applied (for host failure domain)

jarrpa commented:
...OH! I see now. Please add a comment to that effect over the new one you're adding. Specifically mention the function that would modify the first TSC and that having duplicate TSCs is not a problem.

nbalacha commented on pkg/controller/defaults/placements.go:
```diff
@@ -66,7 +66,8 @@ var (
 			TopologySpreadConstraints: []corev1.TopologySpreadConstraint{
-				getTopologySpreadConstraintsSpec(1, "rook-ceph-osd-prepare"),
+				getTopologySpreadConstraintsSpec(1, "rook-ceph-osd-prepare", "rook-ceph-osd"),
+				getTopologySpreadConstraintsSpec(1, "rook-ceph-osd-prepare", "rook-ceph-osd"),
```
I think the reverse approach would be easier to understand - add the extra TSC entry and set it to zone/rack at runtime only if flexible scaling is disabled.

umangachapagain replied:
Agree with @nbalacha 
This looks like a mess to me.

jarrpa replied:
+1

kesavanvt replied:
Thanks,required changes have been done!

nbalacha commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -403,6 +403,10 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster, serverVersion *version.
 						in := getPlacement(sc, "osd-prepare-tsc")
 						(&in).DeepCopyInto(&preparePlacement)
 
+						if sc.Spec.FlexibleScaling == true {
```
Please compare against topologyKey instead of FlexibleScaling because the storageDeviceSets also has a topologyKey.

kesavanvt replied:
Thanks, comparing against the topology key seems to be appropriate and changes are done!

kesavanvt commented:
> ...OH! I see now. Please add a comment to that effect over the new one you're adding. Specifically mention the function that would modify the first TSC and that having duplicate TSCs is not a problem.

I have added the necessary comments. Unfortunately placement with duplicate TSC leads to error ( refer https://kubernetes.io/blog/2020/05/introducing-podtopologyspread/ ). so i have added the statements to remove multi level TSC wherever necessary in case of host failure domain.

umangachapagain commented on pkg/controller/defaults/placements.go:
```diff
@@ -66,7 +66,14 @@ var (
 			},
 			TopologySpreadConstraints: []corev1.TopologySpreadConstraint{
-				getTopologySpreadConstraintsSpec(1, "rook-ceph-osd-prepare"),
+				// getTopologySpreadConstraintsSpec populate values required for topology spread constraints.
```
```suggestion
				// getTopologySpreadConstraintsSpec populates values required for topology spread constraints.
```

umangachapagain commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -403,8 +403,16 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster, serverVersion *version.
 							preparePlacement.TopologySpreadConstraints[0].TopologyKey = topologyKey
+						} else {
+							// Eliminating default multi level topology spread consrtaints.
+							preparePlacement.TopologySpreadConstraints = preparePlacement.TopologySpreadConstraints[:1]
```
If Line 408 executes, what is the value of this line? I don't think this does any change to TSC.

kesavanvt replied:
Now the changes have been made to add multiple TSC only when zone/rack is detected as failure domain.

jarrpa commented:
/test ocs-operator-bundle-e2e-aws

jarrpa reviewed (approved):
Changes look good! A much cleaner and straightforward patch. :)

/lgtm

kesavanvt commented:
/ok-to-test

jarrpa reviewed (approved):
/lgtm
